### DISCUSSION TITLE
Generating resources for SDK-style projects targeting net20/net35

### DISCUSSION THREAD

**Discussion by savornicesei** [dotnet/msbuild#7590] [Q&A] [answered]
Hi all,

I've recently converted an [old project](https://github.com/NAnt2/NAnt2/tree/feature/GH-136) to SDK-style project files since, according to the docs, supports every .NET framework from .NET 2.0 onwards, using Microsoft.NETFramework.ReferenceAssemblies.
There are a couple of resource files with only text resources in them and when building for `TargetFramework = net20` or `net35` it complains about missing ResGen. The error dissapears when I switch to `net40` or later
```
 ResGen.exe not supported on .NET Core MSBuild
```
This happens on both:

- linux (openSUSE 15.3 -wsl,  mono 6.8, msbuild 16.6, .NET SDK 5.0.407)
- windows (win 10, all .NET frameworks, target packs, SDK - including the one for .NET 4.0 and .NET SDK 6.0.202)

There was some work done in this area for .NET Core 3.0 ([#2272](https://github.com/dotnet/msbuild/issues/2272), [#2221](https://github.com/dotnet/msbuild/issues/2221), [docfx#8474](https://github.com/dotnet/sdk/issues/8474), [#4082](https://github.com/dotnet/msbuild/issues/4082)), there is a [GenerateResource](https://docs.microsoft.com/en-us/visualstudio/msbuild/generateresource-task?view=vs-2022) MSBuild task documented in the official docs and there is also a `System.Resources.Extensions` NuGet package that targets .NET 461 and later.

What is the correct way to generate strongly-typed resource files for .NET 2.0/3.5 in SDK-style projects?

Thank you.

**Comment by rainersigwald** [comment, marked as answer]
To target .NET Framework 2.0/3.5 projects that use resources, you must build with `MSBuild.exe`. `dotnet build` is not supported for that scenario.

The reason for this is that .NET Framework resources required using `BinaryFormatter`, which varies between .NET versions, so we cannot create the resources from a .NET 6 process. An improved mechanism is available called `System.Resources.Extensions`, but it supports .NET Framework 4.6.1 and higher, with no plans to support lower framework versions (https://github.com/dotnet/runtime/issues/62282).

However, `MSBuild.exe` retains the ability to use a .NET Framework 3.5 process to create the resources, and works with .NET SDK projects. So to build this project you'll need to use that MSBuild, not the one from the .NET (Core) SDK.

**Reply by savornicesei** [reply]
Thanks @rainersigwald for ~~the explanations~~ complicating my life 😆 

So for anything less than .NET 4.6.2 - MSBuild is required to generate string resources, and not `dotnet build` or `dotnet msbuild` (even though I had the impression the later calls MSBuild under the hood)
Do I need `System.Resources.Extensions` and `<GenerateResourceUsePreserializedResources>true</GenerateResourceUsePreserializedResources>` (according to [dotnet/msbuild#4704](https://github.com/dotnet/msbuild/issues/4704)) for the rest, including netcore / netstandard or just for the rest of .NET framework targets, up to net48?

**Reply by rainersigwald** [reply]
> Thanks @rainersigwald for ~the explanations~ complicating my life 😆

😅

> `dotnet msbuild` (even though I had the impression the later calls MSBuild under the hood)

That is correct, but it uses MSBuild-for-.NET-Core, which doesn't (can't) support features that .NET 6 itself doesn't, so it has some behavior differences from MSBuild-for-.NET-Framework (which is also MSBuild-for-Visual Studio).


> Do I need `System.Resources.Extensions` and `<GenerateResourceUsePreserializedResources>true</GenerateResourceUsePreserializedResources>` (according to [dotnet/msbuild#4704](https://github.com/dotnet/msbuild/issues/4704)) for the rest, including netcore / netstandard or just for the rest of .NET framework targets, up to net48?

You need `System.Resources.Extensions` for any case where

1. You build with `dotnet build`/`dotnet msbuild`, and
2. You have non-string resources, and
3. The target runtime is not .NET Core 3+/.NET 5+.

.NET 6 has the functionality to deserialize resources embedded via `PreserializedResources` built in; .NET Framework and .NET Standard do not and need the DLL.
